docs/development.md

```diff
@@ -0,0 +1,140 @@
+# Parthenon developer guide
+
+### (Kokkos) programming model
+
+The following list contains a few overall design decision that are useful to keep in mind
+
+- Kokkos never hides copies, i.e., there are no hidden deep copies and all data transfer needs to be explicit.
+- Kernel launches (e.g., a parallel region) are non-blocking, i.e., the CPU continues to execute code following a kernel that is executed on a GPU.
+
+### Kokkos wrappers/abstractions
+
+- `par_for` wrappers use inclusive bounds, i.e., the loop will include the last index given
+- `AthenaArrayND` arrays by default allocate on the *device* using default precision configured
+- To create an array on the host with identical layout to the device array either use
+  - `auto arr_host = Kokkos::create_mirror(arr_dev);` to always create a new array even if the device is associated with the host (e.g., OpenMP) or
+  - `auto arr_host = Kokkos::create_mirror_view(arr_dev);` to create an array on the host if the HostSpace != DeviceSpace or get another reference to arr_dev through arr_host if HostSpace == DeviceSpace
+- `par_for` and `Kokkos::deep_copy` by default use the standard stream (on Cuda devices) and are discouraged from use. Use `mb->par_for` and `mb->deep_copy` instead where `mb` is a `MeshBlock` (explanation: each `MeshBlock` has an `ExecutionSpace`, which may be changed at runtime, e.g., to a different stream, and the wrapper within a `MeshBlock` offer transparent access to the parallel region/copy where the `MeshBlock`'s `ExecutionSpace` is automatically used).
+
+An arbitrary-dimensional wrapper for `Kokkos::Views` is available as
+`ParArrayND`. See documentation [here](parthenon_arrays.md).
+
+The wrappers `par_for_outer` and `par_for_inner` provide a nested parallelism interface that is needed for managing memory cached in tightly nested loops. The wrappers are documented [here](nested_par_for.md).
+
+### The need for reductions within function handling `MeshBlock` data
+
+A task (often a function associated with a `MeshBlock`) in the original Athena++ code was guaranteed to only being executed using a single CPU thread at any time.
+Thus, a reduction within a single `MeshBlock`(e.g., finding a minimum value in the `MeshBlock` data or filling a buffer using an incremented `offset` pointer) did not require special care.
+This is *not* true in Parthenon any more as a parallel region may be handled by many (GPU/...) threads in parallel.
+Therefore, a `parallel_reduce` needs to be used instead of a `parallel_for`.
+
+A strong hint where this is in order are places where a single variable is incremented within a kernel or where another reduction over MPI processes follows the preceding computations.
+
+Currently, Parthenon does not provide wrappers for parallel reductions so the raw Kokkos versions are used.
```

Comment:
At some point this would definitely be nice, but given the few places they are currently used I would not put it a the top of the priority list right now.

Reply:
I started an issue so we can keep track of this.